### ISSUE TITLE
Unsafe Two Factor Recovery Codes

### ISSUE THREAD

**Issue by mbeckenbach** [aspnet/Identity#1918]
Hi, 

i've noticed that the two factor recovery codes for authenticator logins are not stored as hashes in AspNetUserTokens table. Shouldn't those be hashed in database like passwords are?

Best regards
Max

**Comment by brockallen** [comment]
This is a dup of this: https://github.com/aspnet/Identity/issues/836, which was closed as part of https://github.com/aspnet/Identity/issues/1341, but come to think of it I don't think they ever added it.

**Comment by mbeckenbach** [comment]
I don’t think this is about general data encryption because of the crazy GDPR. This is more about having unencrypted credentials in the database. It just feels very wrong.

Beste Grüße

Max Beckenbach
Am 11. Aug. 2018, 14:44 +0200 schrieb Brock Allen <notifications@github.com>:

**Comment by brockallen** [comment]
Sure, I agree. That's what the original issue was about.

**Comment by blowdart** [comment]
Arguably this isn't covered by GDPR because it's not personally identifying information (after all, multiple people could get the same codes), so it wasn't part of the GDPR work.

As it'll need some database work it is scheduled for 3.0

**Comment by mbeckenbach** [comment]
I see. My intend was basically to report it just in case someone simply forgot it. I don’t think this is urgent.

I guess the risk of unencrypted codes is not that high as someone would need the users password as well as database access to bypass two factor to get access to someone’s account. If someone manages to get access to your database you have other problems to care about. :)
Am 16. Aug. 2018, 21:10 +0200 schrieb Barry Dorrans <notifications@github.com>:

**Comment by mbeckenbach** [comment]
What does „mfa“ mean? Google said it means „monounsaturated fatty acids“ in german, but i guess thats not what you wanted to said. :-D
Am 16. Aug. 2018, 22:20 +0200 schrieb Brock Allen <notifications@github.com>:

**Comment by brockallen** [comment]
mfa == multi-factor authentication

**Comment by mbeckenbach** [comment]
:) ok.
Am 22. Aug. 2018, 23:38 +0200 schrieb Brock Allen <notifications@github.com>:
